// makepdf/dist/parser/context.js
import { flattenText } from "./inline.js";
class AutoNumberState {
    constructor(_prev, _isAlpha) {
        this._prev = _prev;
        this._isAlpha = _isAlpha;
        this._next = {};
        this._num = 0;
    }
    inc() {
        this._num++;
        this._next = {};
        return this.current();
    }
    current(suffix = "") {
        if (!this._prev)
            return "";
        if (!this._num)
            this._num = 1;
        let cur = this._isAlpha ? String.fromCharCode(64 + this._num) : this._num;
        return this._prev.current(".") + cur + suffix;
    }
    getNextAlpha(name) {
        let id = "A" + (name || "default");
        if (!this._next[id]) {
            this._next[id] = new AutoNumberState(this, true);
        }
        return this._next[id];
    }
    getNextNum(name) {
        let id = "N" + (name || "default");
        if (!this._next[id]) {
            this._next[id] = new AutoNumberState(this);
        }
        return this._next[id];
    }
}
export class OutputContext {
    constructor(config) {
        this.config = config;
        this._numState = new AutoNumberState();
        this._nextHeadingId = 1;
        this._toc = [];
        this._tocTable = {};
        this._refIds = {};
        this._refsToUpdate = [];
    }
    getBaseDir() {
        return this.config.input.baseDir || "./";
    }
    getStyleProps(style) {
        let styles = this.config.styles;
        if (!styles[style]) {
            throw Error("Style not defined: " + style);
        }
        return styles[style];
    }
    getHeadingId() {
        return this._nextHeadingId++;
    }
    getMaxTocLevel() {
        return this.config.output.tocLevel || 1;
    }
    mergeDefinitions(overrides = {}) {
        return { ...this.config.define, ...this.config.output.info, ...overrides };
    }
    addRefId(id, text, headingLevel) {
        text = flattenText(text);
        this._refIds[id] = { text, headingLevel };
        if (headingLevel >= 1 && headingLevel <= this.getMaxTocLevel()) {
            this._toc.push(id);
        }
    }
    addRefToUpdate(id, refNode) {
        this._refsToUpdate.push({ id, refNode });
        return refNode;
    }
    updateRefs() {
        this._refsToUpdate.forEach((r) => {
            if (!this._refIds[r.id])
                throw Error("Reference not found: " + r.id);
            r.refNode.text = this._refIds[r.id]?.text || "";
        });
        Object.assign(this._tocTable, this.config.styles.toc);
        this._tocTable.table = {
            widths: this.config.styles.toc.widths,
            body: this._toc.map((id) => [
                {
                    linkToDestination: id,
                    text: this._refIds[id].text,
                    style: "toc" + this._refIds[id].headingLevel,
                },
                { pageReference: id, alignment: "right" },
            ]),
        };
    }
    autoNumber(pattern) {
        let num = this._numState;
        while (pattern.length) {
            let nextIdx = pattern.indexOf(".") + 1;
            if (pattern[0] === "(") {
                let endIdx = pattern.indexOf(")");
                let name = pattern.slice(1, endIdx > 0 ? endIdx : 1);
                num = num.getNextNum(name);
            }
            else if (pattern[0] === "[") {
                let endIdx = pattern.indexOf("]");
                let name = pattern.slice(1, endIdx > 0 ? endIdx : 1);
                num = num.getNextAlpha(name);
            }
            if (nextIdx <= 0)
                break;
            pattern = pattern.slice(nextIdx);
        }
        return num.inc() + this.config.output.autonumSuffix;
    }
    getTOCTable() {
        return this._tocTable;
    }
}
